tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -117,50 +117,82 @@ impl ServerKey {
-            self.blockshift_right_assign(ct, 1_usize);
+        // rotate left as the blocks are from LSB to MSB
+        ct.blocks.rotate_left(rotations);
+        for block in &mut ct.blocks[num_blocks - rotations..] {
+            self.key.create_trivial_assign(block, 0)
+        }
 
-            // We propagate the last block in order to not lose information
-            self.propagate(ct, ct.blocks.len() - 2);
-            self.blockshift_right_assign(ct, quotient);
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
         }
-    }
 
-    /// Propagates all carries except the last one.
-    /// For development purpose only.
-    fn partial_propagate<PBSOrder: PBSOrderMarker>(&self, ctxt: &mut RadixCiphertext<PBSOrder>) {
-        let len = ctxt.blocks.len() - 1;
-        for i in 0..len {
-            self.propagate(ctxt, i);
+        let message_modulus = self.key.message_modulus.0 as u64;
+        let lut = self
+            .key
+            .generate_accumulator_bivariate(|current_block, mut previous_block| {
+                // left shift not to lose
+                // bits when shifting right afterwards
+                previous_block <<= num_bits_in_block;
+                previous_block >>= shift_within_block;
+
+                // The way of getting carry / message is reversed compared
+                // to the usual way but its normal
+                let message_of_current_block = current_block >> shift_within_block;
+                let carry_of_previous_block = previous_block % message_modulus;
+
+                message_of_current_block + carry_of_previous_block
+            });
+        let partial_blocks = ct.blocks[..num_blocks - rotations]
+            .windows(2)
+            .map(|blocks| {
+                // We are right-shifting, which in our representation
+                // means the previous_block (the one with the carry to be
+                // progatated to the current_block) is the next block in the Vec
+                let (current_block, previous_block) = (&blocks[0], &blocks[1]);
+                self.key
+                    .unchecked_apply_lookup_table_bivariate(current_block, previous_block, &lut)
+            })
+            .collect::<Vec<_>>();
+
+        // We do this block separately as this one does not
+        // need to get incoming bits from it neighbour
+        let lut = self
+            .key
+            .generate_accumulator(|x| (x % message_modulus) >> shift_within_block);
+        self.key
+            .apply_lookup_table_assign(&mut ct.blocks[num_blocks - rotations - 1], &lut);
+
+        // We started with num_blocks, discarded 'rotations' blocks
+        // and did the last one separately
+        let blocks_to_replace = &mut ct.blocks[..num_blocks - rotations - 1];
+        assert_eq!(partial_blocks.len(), blocks_to_replace.len());
+        for (block, shifted_block) in itertools::izip!(blocks_to_replace, partial_blocks) {
```

Comment:
sorry if I was unclear, izip was needed only if it was checking the interator lengths itself, I was fine with the way it was written before, I just found the length check hard to read

btw we have our own izip reimported in core to avoid the itertools dependency

Reply:
ah so what should i do 😁?

Reply:
if you like the izip! look then I woudl use the one hidden in core_crypto

otherwise the previous way was fine

Reply:
ok gonna import the izip from core_crypto